### PR TITLE
[SPARK-59562][ML] Fix stale values in reused multinomial logistic aggregation buffer

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?

This PR initializes every active class slice of the reused multinomial logistic aggregation
buffer with its margin offset before calling GEMM with `beta = 1.0`.

It also adds regression coverage that compares a single block with multiple dense and sparse
blocks. The second block is smaller, and one intercept is zero.

### Why are the changes needed?

`MultinomialLogisticBlockAggregator` reuses its margin buffer across blocks. Previously, slices
whose offset was zero were not cleared. They could therefore retain multipliers produced while
processing the previous block. GEMM then added the next block's margins to those stale values,
producing incorrect loss and gradient results.

The bug was introduced by commit
[`a24dfd697c75`](https://github.com/apache/spark/commit/a24dfd697c75e8fba776bc483b8d63103802c2cb)
(`[SPARK-39114][ML] ml.optim.aggregator avoid re-allocating buffers`), which changed the margin
array from a fresh per-block allocation to a reused buffer while retaining the zero-offset fill
skip that had only been safe for a newly allocated array.

See https://issues.apache.org/jira/browse/SPARK-59562.

### Does this PR introduce _any_ user-facing change?

Yes. Multinomial logistic regression now computes the correct loss and gradient when an
aggregator reuses its buffer across blocks and a class has a zero margin offset.

### How was this patch tested?

Added dense and sparse multi-block regression coverage and ran:

```
JAVA_HOME=/usr/lib/jvm/java-17-openjdk-amd64 ./build/sbt \
  'mllib/testOnly org.apache.spark.ml.optim.aggregator.MultinomialLogisticBlockAggregatorSuite'
```

All 9 tests passed.

### Was this patch authored or co-authored using generative AI tooling?

Generated-by: OpenAI Codex (GPT-5)

### CODE REVIEW

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
The trigger is narrower than all multinomial training, but it is reachable during normal
initialization:

1. Without an initial model, multinomial intercepts are initialized as
   `log1p(classCount) - mean(log1p(classCounts))`.
2. For balanced classes, every intercept is mathematically zero. Floating-point evaluation often
   produces bitwise `0.0`, although it can sometimes leave a very small rounding residual.
3. The initial linear coefficients are zero, so centering does not change these margin offsets.
4. If a partition produces multiple training blocks (the default block size is 1 MB), the first
   block replaces the reusable buffer contents with softmax multipliers. For a bitwise-zero offset,
   the old condition skipped resetting that class slice before the next `gemm(..., beta = 1.0)`, so
   stale multipliers were added to the next block's margins.

After the first optimizer update, offsets are continuous values and an exact zero is generally
unlikely. Thus the bug is configuration-dependent and mostly affects the initial loss/gradient
evaluation of balanced multinomial datasets rather than every iteration or dataset.

**Review by uros-b — APPROVED** [review]

**Comment by uros-b** [comment]
Thank you @zhengruifeng and @srowen!

**Comment by zhengruifeng** [comment]
**Merge Summary:**
- merged into master https://github.com/apache/spark/commit/2ec543b545346bfd01492572d925cf45f489878d
- merged into branch-4.x https://github.com/apache/spark/commit/09177fabc24dee5019b75689ce4d7abe1e5ba892
- merged into branch-4.3 https://github.com/apache/spark/commit/7f68347907120903c73e59b642786a78a2075089
- merged into branch-4.2 https://github.com/apache/spark/commit/6641056fa3205055e48d3775f0fcab713d39dc64
- merged into branch-4.1 https://github.com/apache/spark/commit/12d29b10923787784d244d4192e284797681aa85
- merged into branch-4.0 https://github.com/apache/spark/commit/4d8a64f065adb2b9a6dad299d9139ceae01ebd66

*Posted by `merge_spark_pr.py`*
